source/Ship.cpp

```diff
@@ -2877,24 +3051,68 @@ void Ship::Recharge(bool atSpaceport)
 
 bool Ship::CanRefuel(const Ship &other) const
 {
-	return (fuel - JumpFuel(targetSystem) >= other.JumpFuelMissing());
+	// Can't refuel if current ship has zero fuel capacity.
+	if(attributes.Get("fuel capacity") < 1 || GetSystem() != other.GetSystem() || IsEnemyInEscortSystem())
+		return false;
+	if(CanBeCarried())
+	{
+		// Ensure all escorts have minimum one jump of fuel before refueling all
+		// escorts to maximum.  Only perform "max refueling" from within your
+		// own fleet.
+		if(IsYours() && !other.JumpFuelMissing())
+		{
+			const std::vector<std::weak_ptr<Ship>> allEscorts = (CanBeCarried() && GetParent()) ? GetParent()->GetEscorts() : GetEscorts();
+			bool escortsNeedFuel = false;
+			for(const weak_ptr<Ship> &ptr : allEscorts)
+			{
+				shared_ptr<Ship> escort = ptr.lock();
+				if(!escort || escort->IsParked() || escort->IsDestroyed() || GetSystem() != escort->GetSystem() || (!escort->IsYours() && !escort->GetPersonality().IsEscort()))
+					continue;
+				// Both mission NPC escorts and player-owned escorts should be refueled first.
+				if(escort->JumpFuelMissing())
+					return false;
+				if(escort->IsYours() && escort->IsFuelLow() && !escort->CanBeCarried())
+					escortsNeedFuel = true;
+			}
+			// Prioritize refueling escorts before mission NPCs
+			if(escortsNeedFuel && !other.IsYours() && other.GetPersonality().IsEscort())
+				return false;
+		}
+		else if(!IsYours() && other.IsYours())
+			return false;
+		if(GetParent() == other.shared_from_this())
+			return IsRefueledByRamscoop() && fuel > 25.;
+		else
+			return !IsFuelLow() && other.IsFuelLow() && HasDeployOrder();
+	}
+	return (!IsYours() || other.JumpFuelMissing() || other.GetParent() == this->shared_from_this()) && (fuel - JumpFuel(targetSystem) >= other.JumpFuelMissing());
 }
 
 
 
+// Transfer fuel to another ship.
 double Ship::TransferFuel(double amount, Ship *to)
 {
-	amount = max(fuel - attributes.Get("fuel capacity"), amount);
-	if(to)
```

Comment:
If you remove this sanity check then also make the parameter a reference and verify that the function is never called without a ship.

Reply:
Sorry I think I was confused by your original feedback.  I can add it back in.

Reply:
Fixed